orttraining/orttraining/training_api/checkpoint.h

```diff
@@ -0,0 +1,163 @@
+ */
+template <typename T>
+struct TypedCheckpointProperty : public CheckpointProperty {
+ public:
+  TypedCheckpointProperty(const std::string& prop_name, const T& prop_value)
+      : CheckpointProperty(prop_name), prop_value_(prop_value) {
+  }
+
+  ONNX_NAMESPACE::TensorProto ToTensorProto() override {
+    auto t_proto = ONNX_NAMESPACE::ToTensor<T>(prop_value_);
+
+    // We did not add dims assuming currently we only support scalars.
+    return t_proto;
+  }
+
+  T GetData() const {
+    return prop_value_;
+  }
+
+ private:
+  T prop_value_;
+};
+
+/**
+ * @brief A data class representing traing states.
+ * Including:
+ *     > Parameter states.
+ *     > Optimizer states.
+ *     > User defined training properties, for example 'epoch', 'best_score'.
+ */
+struct CheckpointStates {
+ public:
+  ModuleCheckpointStates module_checkpoint_states;
+  OptimizerCheckpointStates optimizer_checkpoint_states;
+  std::unordered_map<std::string, std::shared_ptr<CheckpointProperty>> named_properties;
+};
+
+PathString CreateFolderIfNotExists(const PathString& path, const std::string& folder_name);
+
+Status CreateOrtValuesFromTensorProtos(
+    const std::vector<const ONNX_NAMESPACE::TensorProto*>& tensor_protos,
+    NameMLValMap& name_to_ort_value);
+
+/**
+ * @brief A single entry for checkpoint utilities.
+ *
+ * A checkpoint is a directory of files:
+ * checkpoint/
+ *   parameters/
+ *       tensors.pbseq - tensor protobuf messages
+ *       tensors.bin - tensor binary data
+ *   optimizers/
+ *       group_0/
+ *           momentum_0/
+ *               tensors.pbseq - tensor protobuf messages
+ *               tensors.bin - tensor binary data
+ *           momentum_1/
+ *               tensors.pbseq - tensor protobuf messages
+ *               tensors.bin - tensor binary data
+ *           properties.pbseq - group-wise property protobuf messages
+ *   properties.pbseq - property protobuf messages
+ */
+struct CheckpointUtils {
+ public:
+  static CheckpointUtils GetInstance() {
+    static CheckpointUtils ckpt;
+    return ckpt;
+  }
+
+  /**
+   * @brief Save ONNX initializers as ORT checkpoint.
+   *
+   * @param tensor_protos trainable parameters in TensorProto format.
+   * @param checkpoint_path folder where checkpoint is saved.
+   * @param model_location onnx model path.
+   * @return Status
+   */
+  static Status SaveORTCheckpoint(const std::vector<const ONNX_NAMESPACE::TensorProto*>& tensor_protos,
+                                  const PathString& checkpoint_path) {
+    ORT_RETURN_IF_ERROR(GetInstance().OrtSaveInternal(tensor_protos, checkpoint_path));
+    return Status::OK();
+  }
+
+  /**
+   * @brief Save training states as ORT checkpoint.
+   *
+   * @param states parameter/optimizer and other user defined training states.
+   * @param checkpoint_path folder where checkpoint is saved.
+   * @return Status
+   */
+  static Status SaveORTCheckpoint(CheckpointStates& states, const PathString& checkpoint_path) {
```

Comment:
In the PR description, "std::vector<const ONNX_NAMESPACE::TensorProto*>" overload is used for the first stage, e.g. "Offline processing calling following SaveORTCheckpoint API to convert those trainable parameters in format of our checkpoint directory."

The idea is, once our offline tools could use call SaveORTCheckpoint (through pybind) to convert "a onnx model" + "trainable parameter names" to the ORT checkpoint format. The reason I proposed to do it in backend, is just to avoid duplication of maintaining ORT checkpoint format in Python frontend and  C++ backend.